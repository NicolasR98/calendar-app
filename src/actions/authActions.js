import { types } from "../types/types";
import { fetchWithoutToken } from "../helpers/fetch";
import Swal from "sweetalert2";

export const startLogin = (email, password) => {
    const auxParams = {
        endpoint: 'auth',
        data: { email, password },
        method: 'POST'
    };

    return async (dispatch) => {
        const resp = await fetchWithoutToken(auxParams);
        const body = await resp.json();

        if (body.ok) {
            const user = {
                uid: body.uid,
                name: body.name,
            };

            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(login(user));
        } else {
            Swal.fire('Error', body.msg, 'error');
        }
    };
};

const login = (user) => ({
    type: types.authLogin,
    payload: user
});