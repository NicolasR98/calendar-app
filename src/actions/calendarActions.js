import { types } from "../types/types";
import { fetchWithToken } from "../helpers/fetch";

export const eventStartAddNew = (event) => {
    return async (dispatch, getState) => {
        const auxParams = {
            endpoint: 'events',
            data: event,
            method: 'POST'
        };
        const { uid, name } = getState().auth;

        try {
            const resp = await fetchWithToken(auxParams);
            const body = await resp.json();

            if (body.ok) {
                event.id = body.event.id;
                event.user = {
                    _id: uid,
                    name,
                };
                dispatch(eventAddNew(event));
            };

        } catch (error) {
            console.error(error);
        }

    };
};

const eventAddNew = (event) => ({
    type: types.calendarAddNew,
    payload: event,
});

export const eventSetActive = (event) => ({
    type: types.calendarSetActive,
    payload: event,
});

export const eventUpdate = (event) => ({
    type: types.calendarUpdate,
    payload: event,
});

export const eventDelete = () => ({ type: types.calendarDelete });

export const eventClearActive = () => ({ type: types.calendarClearActive });