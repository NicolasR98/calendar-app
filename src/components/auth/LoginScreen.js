import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../actions/authActions';
import { useForm } from '../../hooks/useForm';
import './login.css';

export const LoginScreen = () => {
    const dispatch = useDispatch();

    const [formLoginValues, handleLoginInputChange] = useForm({
        loginEmail: '',
        loginPassword: '',
    });
    const [formRegisterValues, handleRegisterInputChange] = useForm({
        registerName: '',
        registerEmail: '',
        registerPassword: '',
        registerPasswordRepeat: '',
    });
    const { loginEmail, loginPassword } = formLoginValues;
    const { registerName, registerEmail, registerPassword, registerPasswordRepeat } = formRegisterValues;

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch(startLogin({
            email: loginEmail,
            password: loginPassword,
        }));
    };

    const handleRegister = (e) => {
        e.preventDefault();

        if (registerPassword !== registerPasswordRepeat) {
            return Swal.fire('Error', 'The passwords must be the same', 'error');
        };

        dispatch(startRegister({
            email: registerEmail,
            password: registerPassword,
            name: registerName,
        }));
    };

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Login</h3>
                    <form onSubmit={handleLogin}>
                        <div className="form-group mb-3">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                name='loginEmail'
                                value={loginEmail}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group mb-3">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                                name='loginPassword'
                                value={loginPassword}
                                onChange={handleLoginInputChange}
                            />
                        </div>
                        <div className="form-group mb-3">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Login"
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Sign Up</h3>
                    <form onSubmit={handleRegister}>
                        <div className="form-group mb-3">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Name"
                                value={registerName}
                                name='registerName'
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group mb-3">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                value={registerEmail}
                                name='registerEmail'
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group mb-3">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                                value={registerPassword}
                                name='registerPassword'
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group mb-3">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repeat password"
                                value={registerPasswordRepeat}
                                name='registerPasswordRepeat'
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group mb-3">
                            <input
                                type="submit"
                                className="btnSubmit"
                                value="Sign Up" />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    );
};