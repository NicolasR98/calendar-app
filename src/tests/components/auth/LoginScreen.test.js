import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';

import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/authActions';


jest.mock('../../../actions/authActions', () => ({
    startLogin: jest.fn(),
    startRegister: jest.fn(),
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};
const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <LoginScreen />
    </Provider>
);


describe('Tests on <LoginScreen />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('Should display correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Should call the dispatch of the login action', () => {
        const emailEvent = {
            target: {
                name: 'loginEmail',
                value: 'test@test.com',
            },
        };
        const passwordEvent = {
            target: {
                name: 'loginPassword',
                value: '123456'
            },
        };
        const expectedArgs = {
            email: emailEvent.target.value,
            password: passwordEvent.target.value,
        };

        // Simulate user credential input
        wrapper.find('input[name="loginEmail"]').simulate('change', emailEvent);
        wrapper.find('input[name="loginPassword"]').simulate('change', passwordEvent);
        const loginForm = wrapper.find('form').at(0);

        loginForm.prop('onSubmit')({
            preventDefault() { }
        });

        expect(startLogin).toHaveBeenCalledWith(expectedArgs);
    });

    test('Should not do a register if the password do not match', () => {
        const nameEvent = {
            target: {
                name: 'registerName',
                value: 'testname',
            },
        };
        const emailEvent = {
            target: {
                name: 'registerEmail',
                value: 'test@test.com',
            },
        };
        const passwordEvent = {
            target: {
                name: 'registerPassword',
                value: '123456'
            },
        };
        const passwordRepeatEvent = {
            target: {
                name: 'registerPasswordRepeat',
                value: 'another_password'
            },
        };

        const expectedSwalError = ['Error', 'The passwords must be the same', 'error'];

        // Simulate user credential input
        wrapper.find('input[name="registerName"]').simulate('change', nameEvent);
        wrapper.find('input[name="registerEmail"]').simulate('change', emailEvent);
        wrapper.find('input[name="registerPassword"]').simulate('change', passwordEvent);
        wrapper.find('input[name="registerPasswordRepeat"]').simulate('change', passwordRepeatEvent);
        const registerForm = wrapper.find('form').at(1);

        registerForm.prop('onSubmit')({
            preventDefault() { }
        });

        expect(Swal.fire).toHaveBeenCalledWith(...expectedSwalError);
        expect(startRegister).not.toHaveBeenCalled();
    });
    test('Should do a register if passwords matches', () => {
        const nameEvent = {
            target: {
                name: 'registerName',
                value: 'testname',
            },
        };
        const emailEvent = {
            target: {
                name: 'registerEmail',
                value: 'test@test.com',
            },
        };
        const passwordEvent = {
            target: {
                name: 'registerPassword',
                value: '123456'
            },
        };
        const passwordRepeatEvent = {
            target: {
                name: 'registerPasswordRepeat',
                value: '123456'
            },
        };

        // Simulate user credential input
        wrapper.find('input[name="registerName"]').simulate('change', nameEvent);
        wrapper.find('input[name="registerEmail"]').simulate('change', emailEvent);
        wrapper.find('input[name="registerPassword"]').simulate('change', passwordEvent);
        wrapper.find('input[name="registerPasswordRepeat"]').simulate('change', passwordRepeatEvent);
        const registerForm = wrapper.find('form').at(1);

        registerForm.prop('onSubmit')({
            preventDefault() { }
        });

        expect(Swal.fire).not.toHaveBeenCalled();
        expect(startRegister).toHaveBeenCalled();
    });
});