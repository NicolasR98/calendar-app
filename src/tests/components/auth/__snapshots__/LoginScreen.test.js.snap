// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <LoginScreen /> Should display correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Login
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Sign Up
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="form-control"
            name="registerPasswordRepeat"
            onChange={[Function]}
            placeholder="Repeat password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group mb-3"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Sign Up"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
